### PR TITLE
Add Ternary Search algorithm for optimization and array searching [HACKTOBERFEST 2025]

### PR DESCRIPTION
**Pull request by piyushkumar0707**
## 🚀 Overview
This PR implements the **Ternary Search** algorithm - a divide-and-conquer technique that extends beyond simple searching to solve optimization problems on unimodal functions.

## ✨ Features
- ✅ **Unimodal function optimization** - Find maximum/minimum of single-peaked functions
- ✅ **Array searching** - Alternative to binary search with 3-way division
- ✅ **Recursive & iterative implementations** - Both approaches for educational comparison
- ✅ **Real-world optimization examples** - Production cost minimization
- ✅ **Performance comparison** - Benchmarks against binary search
- ✅ **Comprehensive testing** - Edge cases and various function types

## 🎯 Why This Matters
Ternary Search is essential for:
- **Mathematical optimization** - Finding optimal solutions in unimodal functions
- **Competitive programming** - Common technique in programming contests
- **Engineering applications** - Cost minimization, resource optimization
- **Machine learning** - Hyperparameter tuning for unimodal loss functions
- **Scientific computing** - Finding extrema in experimental data
- **Algorithm education** - Understanding divide-and-conquer variants

## 📚 Implementation Details
- **Time Complexity:** O(log₃ n) ≈ O(log n) for both searching and optimization
- **Space Complexity:** O(1) iterative, O(log n) recursive
- **Algorithm:** Divides search space into three equal parts instead of two
- **Precision Control:** Configurable precision for optimization problems

## 🔧 Core Functions
- **`ternary_search_maximum()`** - Find maximum of unimodal functions
- **`ternary_search_minimum()`** - Find minimum of unimodal functions
- **`ternary_search_array_recursive()`** - Recursive array searching
- **`ternary_search_array_iterative()`** - Iterative array searching
- **`ternary_search_with_count()`** - Performance analysis version

## 🧪 Comprehensive Examples
- **Quadratic function optimization** - Finding vertex of parabolas
- **Array searching** - Comparison with binary search performance
- **Real optimization problem** - Production cost minimization
- **Edge cases** - Empty arrays, single elements, not found scenarios
- **Performance benchmarking** - Comparison counts with binary search

## 📊 Algorithm Comparison
**Ternary vs Binary Search:**
- Ternary: ~2 comparisons per iteration, reduces to 1/3 each time
- Binary: ~1 comparison per iteration, reduces to 1/2 each time
- **Trade-off:** More comparisons per iteration but faster convergence

## 🎯 Optimization Applications
**Production Cost Example:**
```r
cost_function <- function(x) 100 + 2*x + 0.001*x^2 - 0.1*x*log(x + 1)
optimal_quantity <- ternary_search_minimum(cost_function, 1, 1000)

### CODE REVIEW

**Comment by siriak** [comment]
Please check comments

**Comment by piyushkumar0707** [comment]
@siriak  all conflicts resolved

File: searches/ternary_search.r [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    
+    if (target < arr[mid1]) {
+      right <- mid1 - 1
+    } else if (target > arr[mid2]) {
+      left <- mid2 + 1
+    } else {
+      left <- mid1 + 1
+      right <- mid2 - 1
+    }
+  }
```
[the remark this answers is not in the document]
Comment (siriak) [reply searches/ternary_search.r:200]: Is this comment addressed?

[the remark this answers is not in the document]
Comment (siriak) [reply searches/ternary_search.r:200]: You are just resolving the conversation without addressing the feedback and this has already happened twice. This violates our code of conduct, specifically:
```
Other conduct which could reasonably be considered inappropriate in a professional setting
```
Therefore I am closing the pull request and labeling it accordingly.
